src/zalsa_local.rs

```diff
@@ -339,12 +314,36 @@ pub(crate) struct QueryRevisions {
     pub(super) tracked_struct_ids: IdentityMap,
 
     pub(super) accumulated: Option<Box<AccumulatedMap>>,
+
     /// [`InputAccumulatedValues::Empty`] if any input read during the query's execution
     /// has any direct or indirect accumulated values.
     pub(super) accumulated_inputs: AtomicCell<InputAccumulatedValues>,
+
+    /// This result was computed based on provisional values from
+    /// these cycle heads. The "cycle head" is the query responsible
+    /// for managing a fixpoint iteration. In a cycle like
+    /// `--> A --> B --> C --> A`, the cycle head is query `A`: it is
+    /// the query whose value is requested while it is executing,
+    /// which must provide the initial provisional value and decide,
+    /// after each iteration, whether the cycle has converged or must
+    /// iterate again.
+    pub(super) cycle_heads: FxHashSet<DatabaseKeyIndex>,
```

Comment:
I think we should consider using a `Option<Box<FxHashSet<..>>>` here similar to `accumulated (https://github.com/salsa-rs/salsa/pull/636). Doing the same for accumulated helped improve performance by a few percent and this might be part of the reason why we see an overall regression.
```suggestion
    pub(super) cycle_heads: Option<Box<FxHashSet<DatabaseKeyIndex>>>,
```

It may also be worth considering to create a `CycleHeads` new type wrapper around `Option<Box<FxHashSet>>`

Reply:
Done